src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -370,9 +370,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     private def errorNotStable(tpt: Tree, found: Type) = { TypeNotAStablePrefixError(tpt, found); false }
 
     /** Check that `tpt` refers to a non-refinement class type */
-    def checkClassType(tpt: Tree): Boolean = {
+    def checkClassOrModuleType(tpt: Tree): Boolean = {
       val tpe = unwrapToClass(tpt.tpe)
-      isNonRefinementClassType(tpe) || errorNotClass(tpt, tpe)
+      isNonRefinementClassType(tpe, allowModules = true) || errorNotClass(tpt, tpe)
```

Comment:
Does everything pass (the new test and the `classOf` test) by subsuming the `case SingleType(_, sym) if sym.isModule => true` additional case here?  Seems cleaner and more local than "generalising" a boolean function with a new boolean parameter.